<a href="https://colab.research.google.com/github/sulaimanbehzad/Clustering_according_to_item_popularity/blob/main/clustering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Clustering of Digikala dataset based on items popularity  


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder, normalize

### Cloning the github repository in colab

In [2]:
% cd /content/drive/MyDrive/Clustering_according_to_item_popularity/


/content/drive/MyDrive/Clustering_according_to_item_popularity


In [3]:
# ! ls

In [4]:
# ! git clone https://github.com/sulaimanbehzad/Clustering_according_to_item_popularity.git

In [5]:
! git status

On branch main
Your branch and 'origin/main' have diverged,
and have 2 and 5 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	Clustering_according_to_item_popularity/
	preprocessed_item_recomm.csv

nothing added to commit but untracked files present (use "git add" to track)


In [6]:
! git pull


*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@ab657937bb61.(none)')


In [7]:
# ! git push -u origin main

In [8]:
# ! git commit -m 'redundant files removed'

### Reading and preprocessing the dataset

In [9]:
df = pd.read_excel('/content/drive/MyDrive/Clustering_according_to_item_popularity/keifiat.xlsx')

In [10]:
df.head()

product_id  ...                                      disadvantages
0        3692  ...                                                NaN
1       90213  ...                                          ["ندارد"]
2       59473  ...                                                NaN
3      120499  ...  ["کیفیت پایین اسپیکر\r","حاشیه خالی زیر صفحه ن...
4       67200  ...                                    ["ظرفیت متوسط"]

[5 rows x 12 columns]

In [11]:
df.describe()

product_id       user_id          likes       dislikes
count  100000.000000  1.000000e+05  100000.000000  100000.000000
mean   426482.510140  4.642959e+06       3.114790       1.649460
std    232470.511159  2.619034e+06       8.050031       8.912705
min       300.000000  4.661320e+05       0.000000       0.000000
25%    220871.250000  2.613059e+06       0.000000       0.000000
50%    427858.000000  4.710290e+06       1.000000       0.000000
75%    619397.250000  7.094806e+06       4.000000       1.000000
max    909423.000000  8.623288e+06     854.000000    1344.000000

Select the attributes that will give information about the popularity of the item

In [12]:
df1 = df[['product_id', 'product_title','likes', 'dislikes','comment', 'recommend']]

Check the dataframe for null values and handle them

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   product_id     100000 non-null  int64 
 1   product_title  100000 non-null  object
 2   likes          100000 non-null  int64 
 3   dislikes       100000 non-null  int64 
 4   comment        99883 non-null   object
 5   recommend      100000 non-null  object
dtypes: int64(3), object(3)
memory usage: 4.6+ MB


In [14]:
df1.isnull().sum()

product_id         0
product_title      0
likes              0
dislikes           0
comment          117
recommend          0
dtype: int64

looking at the database 'recommend' we can see that \N values also represent null values which wouldn't be considered in the above

In [15]:
df1['recommend'].value_counts()

recommended        36972
\N                 36382
not_recommended    16110
no_idea            10536
Name: recommend, dtype: int64

we can see that beside the 117 missing values we have 36382 records with \N values  
now we understand tha \N too should be considered as missing value


In [16]:
missing_values = ["\\N", "na", "--", "NA", "n/a"]
df = pd.read_excel('/content/drive/MyDrive/Clustering_according_to_item_popularity/keifiat.xlsx', na_values = missing_values)
df1 = df[['product_id', 'product_title', 'likes', 'dislikes','comment', 'recommend']]


after importing database with newly found aspects on NA we are going to handle null values

In [17]:
df1.isnull().sum()

product_id           0
product_title        0
likes                0
dislikes             0
comment            118
recommend        36382
dtype: int64

In [18]:
df1['recommend'].value_counts()

recommended        36972
not_recommended    16110
no_idea            10536
Name: recommend, dtype: int64

In [19]:
most_common = df1["recommend"].value_counts().idxmax
# df1['recommend'].fillna(median, inplace=True)
most_common

<bound method Series.idxmax of recommended        36972
not_recommended    16110
no_idea            10536
Name: recommend, dtype: int64>

replacing the null values with one of the above would falsely impact the clustering so let's take another approach:  
1. use interpolation to find values
2. fill with mean/median

in order to use interpolation let us first use label encoder to obtain numerical values

In [20]:
df1['recommend'].fillna(value='null_values', inplace=True)
df1.isnull().sum()

/usr/local/lib/python3.6/dist-packages/pandas/core/series.py:4536: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


product_id         0
product_title      0
likes              0
dislikes           0
comment          118
recommend          0
dtype: int64

In [21]:
le = LabelEncoder()
le.fit(df1['recommend'])
le.classes_

array(['no_idea', 'not_recommended', 'null_values', 'recommended'],
      dtype=object)

In [22]:
recomm_trans = le.transform(df1['recommend'])
recomm_trans

array([2, 3, 1, ..., 3, 0, 0])

In [23]:
df1['recommend'] = recomm_trans

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [24]:
df1[['recommend']]

recommend
0              2
1              3
2              1
3              0
4              0
...          ...
99995          3
99996          2
99997          3
99998          0
99999          0

[100000 rows x 1 columns]

In [25]:
df1['recommend'] = df1['recommend'].replace(2,np.nan)
df1['recommend']

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0        NaN
1        3.0
2        1.0
3        0.0
4        0.0
        ... 
99995    3.0
99996    NaN
99997    3.0
99998    0.0
99999    0.0
Name: recommend, Length: 100000, dtype: float64

In [26]:
recomm_filled = df1['recommend'].interpolate(method='linear')
recomm_filled.isnull().sum()

1

In [27]:
df1['recommend'] = recomm_filled
df1.head(50)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


product_id  ... recommend
0         3692  ...       NaN
1        90213  ...  3.000000
2        59473  ...  1.000000
3       120499  ...  0.000000
4        67200  ...  0.000000
5       133722  ...  3.000000
6       148509  ...  3.000000
7        56871  ...  2.000000
8        49738  ...  1.000000
9       161548  ...  0.000000
10        5107  ...  3.000000
11      156041  ...  3.000000
12      134367  ...  3.000000
13      114006  ...  3.000000
14      151402  ...  3.000000
15      146879  ...  3.000000
16       81254  ...  2.000000
17       26368  ...  1.000000
18       90894  ...  0.000000
19       21340  ...  0.500000
20      144139  ...  1.000000
21      138040  ...  1.666667
22      105719  ...  2.333333
23      147165  ...  3.000000
24      116387  ...  3.000000
25       73064  ...  3.000000
26      110040  ...  3.000000
27      139131  ...  3.000000
28       80854  ...  3.000000
29      163519  ...  3.000000
30      102068  ...  3.000000
31      129988  ...  1.000000
32      130386  ...  3.000000
33      160008  ...  3.000000
34       93850  ...  3.000000
35      154624  ...  3.000000
36      106471  ...  3.000000
37      112714  ...  3.000000
38      104427  ...  3.000000
39      133709  ...  3.000000
40       55495  ...  3.000000
41      121304  ...  3.000000
42      150718  ...  3.000000
43      154559  ...  3.000000
44      127328  ...  3.000000
45      162788  ...  3.000000
46       24666  ...  3.000000
47       81429  ...  3.000000
48       58800  ...  3.000000
49      149656  ...  3.000000

[50 rows x 6 columns]

In [28]:
df1.isnull().sum()

product_id         0
product_title      0
likes              0
dislikes           0
comment          118
recommend          1
dtype: int64

In [29]:
df1.dropna(inplace=True)
df1.isnull().sum()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


product_id       0
product_title    0
likes            0
dislikes         0
comment          0
recommend        0
dtype: int64

## handling NULL values is DONE!!! 

In [30]:
df1.to_csv('preprocessed_item_recomm.csv',index=False, encoding='utf-8-sig')

In [64]:
prep_df = pd.read_csv('/content/drive/MyDrive/Clustering_according_to_item_popularity/preprocessed_item_recomm.csv')
prep_df['recommend']=prep_df['recommend'].round()
prep_df['recommend'] = pd.to_numeric(prep_df['recommend'], downcast='integer')
inv_tr=le.inverse_transform(prep_df['recommend'])
inv_tr

array(['recommended', 'not_recommended', 'no_idea', ..., 'recommended',
       'no_idea', 'no_idea'], dtype=object)

In [65]:
prep_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99881 entries, 0 to 99880
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     99881 non-null  int64 
 1   product_title  99881 non-null  object
 2   likes          99881 non-null  int64 
 3   dislikes       99881 non-null  int64 
 4   comment        99881 non-null  object
 5   recommend      99881 non-null  int8  
dtypes: int64(3), int8(1), object(2)
memory usage: 3.9+ MB


In [66]:
dummies = pd.get_dummies(prep_df['recommend'])
dummies.head()

0  1  2  3
0  0  0  0  1
1  0  1  0  0
2  1  0  0  0
3  1  0  0  0
4  0  0  0  1

recommended: 3  
not_recommended: 1  
no_idea: 0 

In [67]:
prep_df['recommended'] = dummies.iloc[:, 3]
prep_df['not_recommended'] = dummies.iloc[:, 1]
prep_df['no_idea'] = dummies.iloc[:, 0]
prep_df.drop(columns=['recommend'], inplace=True)
prep_df.head(10)

product_id  ... no_idea
0       90213  ...       0
1       59473  ...       0
2      120499  ...       1
3       67200  ...       1
4      133722  ...       0
5      148509  ...       0
6       56871  ...       0
7       49738  ...       0
8      161548  ...       1
9        5107  ...       0

[10 rows x 8 columns]

Here we aim to obtain sum of number times an item has been recommended 

In [68]:
sum_of_recomm = prep_df.groupby(by=['product_id'], as_index=False).sum()
sum_of_recomm.head(10)


product_id  likes  dislikes  recommended  not_recommended  no_idea
0         300      5         1            2                1        0
1         310      2         0            1                0        0
2         352      0         0            0                0        1
3         363      0         0            1                0        0
4         423      0         0            1                0        0
5         481      1         0            0                0        0
6         482      1         1            1                0        0
7         485      0         0            0                0        0
8         496      0         0            0                0        1
9         515      9         0            1                0        0

Drop duplicates here



In [69]:
unique_products = prep_df.filter(['product_id', 'product_title'])
unique_products.drop_duplicates(subset='product_id', inplace=True, keep='first')
unique_products.head()

product_id                                      product_title
0       90213  شارژر همراه شیاومی مدل NDY-02-AN با ظرفیت 1000...
1       59473              یدک پولیشر میکروفایبر مهسان مدل 20119
2      120499   گوشی موبایل هوآوی آنر مدل 5X KIW-L21 دو سیم‌کارت
3       67200  شارژر همراه شیائومی مدل Mi ظرفیت 5000 میلی آمپ...
4      133722               لپ تاپ 15 اینچی ایسوس مدل N501VW - A

In [70]:
unique_products.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50189 entries, 0 to 99879
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     50189 non-null  int64 
 1   product_title  50189 non-null  object
dtypes: int64(1), object(1)
memory usage: 1.1+ MB


More than 49000 products with duplicate records were deleted as a result
let's save the preproccessed dataframe upto now

In [71]:
final_df = pd.merge(sum_of_recomm,unique_products,on="product_id")
final_df.head()

product_id  likes  ...  no_idea                          product_title
0         300      5  ...        0  پرینتر لیزری اچ پی مدل LaserJet P2035
1         310      2  ...        0               اچ پی لیزر جت پی 2055 دی
2         352      0  ...        1   گوشی موبایل نوکیا 5800 نویگیشن ادیشن
3         363      0  ...        0          گوشی موبایل نوکیا 6303 کلاسیک
4         423      0  ...        0          لپ تاپ ایسر اسپایر 5738 زد جی

[5 rows x 7 columns]

In [73]:
final_df.head(50)

product_id  ...                                      product_title
0          300  ...              پرینتر لیزری اچ پی مدل LaserJet P2035
1          310  ...                           اچ پی لیزر جت پی 2055 دی
2          352  ...               گوشی موبایل نوکیا 5800 نویگیشن ادیشن
3          363  ...                      گوشی موبایل نوکیا 6303 کلاسیک
4          423  ...                      لپ تاپ ایسر اسپایر 5738 زد جی
5          481  ...                                  سونی پلی استیشن 2
6          482  ...          سونی پلی استیشن پورتابل (پی اس پی) - 3000
7          485  ...                      گوشی موبایل نوکیا 6500 اسلاید
8          496  ...                                         نینتندو وی
9          515  ...                            گوشی موبایل نوکیا ان 82
10         542  ...                             گوشی موبایل نوکیا 1202
11         553  ...                             گوشی موبایل نوکیا 5700
12         623  ...             گوشی موبایل سونی اریکسون کا 970 - اِلم
13         696  ...                    گوشی موبایل سونی اریکسون کا 810
14         882  ...  هارد وسترن دیجیتال پرتابل مای پاسپورت استدیو -...
15         935  ...           دوربین فیلمبرداری سونی اچ دی آر-اس آر 11
16         953  ...            مانیتور سامسونگ مدل 2033SN سایز 20 اینچ
17         965  ...       دوربین دیجیتال کانن پاورشات اس ایکس 20 آی اس
18         996  ...          ماوس باسیم فراسو مدل FOM-1380 با رابط USB
19        1179  ...   پرینتر مخصوص چاپ عکس اپسون مدل Stylus Photo 1410
20        1204  ...                             لپ تاپ دل وسترو 1015-D
21        1317  ...                             گوشی موبایل نوکیا 1203
22        1334  ...                     اسپیکر کریتیو اینسپایر ام 4500
23        1336  ...                          اسپیکر کریتیو مدل SBS A60
24        1339  ...                             اسپیکر ادیفایر ام 3310
25        1345  ...                        لپ تاپ دل اینسپایرون 1564-C
26        1464  ...                    کریتیو زن استایل 300 8 گیگابایت
27        1482  ...              پرینتر لیزری اچ پی مدل LaserJet P1102
28        1540  ...                              پرزنتر رادیویی بی سیم
29        2097  ...             اپل آی پاد تاچ نسل چهارم - 32 گیگابایت
30        2118  ...                              اچ پی لیزر جت ام 1132
31        2179  ...                                    اپل آی پاد شافل
32        2306  ...                         کیبورد فراسو اف سی آر 8910
33        2535  ...       فلش مموری ای دیتا مدل C906 ظرفیت 16 گیگابایت
34        2582  ...               گوشی موبایل سونی اریکسون اکسپریا پلی
35        2640  ...  مودم روتر +ADSL2 بی‌ سیم N300 تی پی-لینک مدل T...
36        2739  ...              کارت شبکه بی‌سیم تی پی-لینک TL-WN821N
37        2747  ...    کارت شبکه USB و بی‌سیم تی پی-لینک مدل TL-WN722N
38        2767  ...             ماوس مایکروسافت مدل Arc Touch رنگ مشکی
39        2842  ...                        گوشی موبایل نوکیا ایکس 7-00
40        2847  ...            کیبورد فراسو FCR-6160 USB با حروف فارسی
41        2854  ...  ماوس باسیم و اپتیکال فراسو مدل FOM-3505  با را...
42        2924  ...                          لپ تاپ دل اینسپایرون 5110
43        2953  ...                 هدست روگوشی استریو لاجیتک مدل H110
44        3159  ...            گوشی موبایل سامسونگ آی 997 اینفیوز 4 جی
45        3261  ...   پرینتر چندکاره لیزری اچ پی مدل LaserJet M1536DNF
46        3275  ...  دوربین دیجیتال سونی سایبرشات دی اس سی - دبلیو ...
47        3281  ...   کیبورد و ماوس بی‌سیم مایکروسافت مدل Desktop 3000
48        3288  ...           کیبورد و ماوس مایکروسافت مدل Desktop 600
49        3290  ...              کیبورد مایکروسافت بولوتوث موبایل 6000

[50 rows x 7 columns]